python/ray/ml/examples/tfx_tabular_train_to_serve.py

```diff
@@ -0,0 +1,201 @@
+from fastapi import Request
+import requests
+from sklearn.model_selection import train_test_split
+import tensorflow as tf
+
+import ray
+from ray import serve
+from ray import train
+from ray.ml.checkpoint import Checkpoint
+from ray.ml.predictors.integrations.tensorflow import TensorflowPredictor
+from ray.ml.train.integrations.tensorflow import TensorflowTrainer
+from ray.ml.preprocessors import (
+    BatchMapper,
+    Chain,
+    OneHotEncoder,
+    SimpleImputer,
+)
+from ray.serve.model_wrappers import ModelWrapperDeployment
+from ray.train.tensorflow import prepare_dataset_shard
+
+
+taxi_data = pd.read_csv(
+    "https://raw.githubusercontent.com/tensorflow/tfx/master/"
+    "tfx/examples/chicago_taxi_pipeline/data/simple/data.csv"
+)
+taxi_data["is_big_tip"] = taxi_data["tips"] / taxi_data["fare"] > 0.2
+# Remove the following features as we don't have bucketizer yet.
+taxi_data = taxi_data.drop(
+    [
+        "tips",
+        "fare",
+        "dropoff_latitude",
+        "dropoff_longitude",
+        "pickup_latitude",
+        "pickup_longitude",
+        "pickup_census_tract",
+    ],
+    axis=1,
+)
+
+# `from_csv` doesn't work right now.
+train_data, test_data = train_test_split(
+    taxi_data, stratify=taxi_data[["is_big_tip"]], random_state=1113
+)
+train_dataset = ray.data.from_pandas(train_data)
+test_data_label = test_data["is_big_tip"].values
+test_data = test_data.drop(["is_big_tip"], axis=1)
+
+
+imputer1 = SimpleImputer(
+    ["dropoff_census_tract"], strategy="constant", fill_value=17031839100
+)
+imputer2 = SimpleImputer(
+    ["pickup_community_area", "dropoff_community_area"],
+    strategy="constant",
+    fill_value=8,
+)
+imputer3 = SimpleImputer(["payment_type"], strategy="constant", fill_value="Cash")
+imputer4 = SimpleImputer(
+    ["company"], strategy="constant", fill_value="Taxi Affiliation Services"
+)
+imputer5 = SimpleImputer(
+    ["trip_start_timestamp", "trip_miles", "trip_seconds"], strategy="mean"
+)
+
+ohe = OneHotEncoder(
+    columns=[
+        "trip_start_hour",
+        "trip_start_day",
+        "trip_start_month",
+        "dropoff_census_tract",
+        "pickup_community_area",
+        "dropoff_community_area",
+        "payment_type",
+        "company",
+    ],
+    limit={
+        "dropoff_census_tract": 25,
+        "pickup_community_area": 20,
+        "dropoff_community_area": 20,
+        "payment_type": 2,
+        "company": 7,
+    },
+)
+
+
+def fn(df):
+    df["trip_start_year"] = pd.to_datetime(df["trip_start_timestamp"], unit="s").dt.year
+    return df
+
+
+chained_pp = Chain(
+    imputer1,
+    imputer2,
+    imputer3,
+    imputer4,
+    imputer5,
+    ohe,
+    BatchMapper(fn),
+    BatchMapper(lambda x: x.drop(["trip_start_timestamp"], axis=1)),
+)
+
+# This is came up by manually inspecting the dataset after preprocessing.
+# Hopefully we can get rid of it in the future.
+INPUT_SIZE = 120
+# Not sure what to do with partial batch.
+BATCH_SIZE = 1
```

Comment:
@amogkam @Yard1 
Do you guys know how partial batch is dealt with?

Reply:
From `tf.data.Dataset.from_generator`, which we use in `to_tf`:

```
    The components of the resulting element will have an additional outer
    dimension, which will be `batch_size` (or `N % batch_size` for the last
    element if `batch_size` does not divide the number of input elements `N`
    evenly and `drop_remainder` is `False`). If your program depends on the
    batches having the same outer dimension, you should set the `drop_remainder`
    argument to `True` to prevent the smaller batch from being produced.
```

Reply:
I don't think we expose this argument, or provide our own implementation of `drop_last`, though.

Reply:
Got it!

if I set BATCH_SIZE=8,
I got
```
ray.exceptions.RayTaskError(InvalidArgumentError): ray::BaseWorkerMixin._BaseWorkerMixin__execute() (pid=58890, ip=127.0.0.1, repr=<ray.train.worker_group.BaseWorkerMixin object at 0x7fd6603c9490>)
  File "/Users/xwjiang/ray/python/ray/train/worker_group.py", line 26, in __execute
    return func(*args, **kwargs)
  File "/Users/xwjiang/ray/python/ray/train/backend.py", line 489, in end_training
    output = session.finish()
  File "/Users/xwjiang/ray/python/ray/train/session.py", line 118, in finish
    func_output = self.training_thread.join()
  File "/Users/xwjiang/ray/python/ray/train/utils.py", line 96, in join
    raise self.exc
  File "/Users/xwjiang/ray/python/ray/train/utils.py", line 89, in run
    self.ret = self._target(*self._args, **self._kwargs)
  File "/Users/xwjiang/ray/python/ray/train/utils.py", line 138, in <lambda>
    return lambda: train_func(config)
  File "python/ray/ml/examples/tfx_tabular_train_to_serve.py", line 163, in train_loop_per_worker
    model.fit(tf_dataset)
  File "/Users/xwjiang/anaconda3/envs/ray2/lib/python3.7/site-packages/keras/utils/traceback_utils.py", line 67, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/Users/xwjiang/anaconda3/envs/ray2/lib/python3.7/site-packages/tensorflow/python/eager/execute.py", line 55, in quick_execute
    inputs, attrs, num_outputs)
tensorflow.python.framework.errors_impl.InvalidArgumentError: Graph execution error:

TypeError: `generator` yielded an element of shape (3, 120) where an element of shape (8, 120) was expected.
ray::BaseWorkerMixin._BaseWorkerMixin__execute() (pid=58890, ip=127.0.0.1, repr=<ray.train.worker_group.BaseWorkerMixin object at 0x7fd6603c9490>)

  File "/Users/xwjiang/anaconda3/envs/ray2/lib/python3.7/site-packages/tensorflow/python/ops/script_ops.py", line 271, in __call__
    ret = func(*args)

  File "/Users/xwjiang/anaconda3/envs/ray2/lib/python3.7/site-packages/tensorflow/python/autograph/impl/api.py", line 642, in wrapper
    return func(*args, **kwargs)

  File "/Users/xwjiang/anaconda3/envs/ray2/lib/python3.7/site-packages/tensorflow/python/data/ops/dataset_ops.py", line 1048, in generator_py_func
    f"`generator` yielded an element of shape {ret_array.shape} "

TypeError: `generator` yielded an element of shape (3, 120) where an element of shape (8, 120) was expected.
```

How should I proceed here?

Reply:
also @clarkzinzow for `to_tf` questions.

Reply:
yeah we probably need to add this functionality into to_tf

Reply:
@jianoaix

Reply:
Datasets supports dropping (or keeping, the default) partial batches via the `drop_last` argument for `ds.iter_batches()`: https://github.com/ray-project/ray/blob/85e6b62acdfc7efa0106d1a5d3d8df72c7a43806/python/ray/data/dataset.py#L2075

This is exposed in `.to_torch()` already: https://github.com/ray-project/ray/blob/85e6b62acdfc7efa0106d1a5d3d8df72c7a43806/python/ray/data/dataset.py#L2129

So do we just need to expose this for `.to_tf()`?

Reply:
@clarkzinzow yep, we already do that for `to_torch`

Reply:
Nice, looks like a small change?

Reply:
Let me make a PR real quick

Reply:
https://github.com/ray-project/ray/pull/24666